### PR TITLE
Update FreeBSD installation docs for LLDB support

### LINKED ISSUE #797: FreeBSD port team needs to document LLDB-setup
We should either have a documented (version-controlled?) script/gist which can be used (and re-used) to get the correct LLDB-version built and installed, or we need to get an official package for this which can be installed like any of our other dependencies.

While we can rely on external wikis and end-user copy-pasting, I would rather have something more closely tied to this project for increased traceability.

Getting a simple and repeatable LLDB installation will not only help getting work done(tm), but might also assist [getting a complete CI-build](https://github.com/dotnet/coreclr/issues/793): In lack of LLDB, we're currently we're not the SOS LLDB-plugin.

As for LLDB version... Should we aim for 3.5? 3.6? or latest? Does it matter?

I think Linux settled for LLDB 3.6, despite building with LLVM 3.5 because LLDB 3.5 (from official packages) was "buggy" on Ubuntu, but I don't know the full back-story for this. Are there any similar concerns which should govern which version we use for FreeBSD?

@emaste : You were briefly involved with getting LLDB into the pkg-repository. Any progress or news since last time? If you've made any progress, what version(s) are you currently working with?

### PR DESCRIPTION
**Pull request by tony**
In short, previously LLDB support for FreeBSD and CoreCLR was a hit and miss thing on some systems. It would at the very least require building LLVM, Clang from the ground up to get LLDB.  The new ``devel/llvm37`` and ``devel/llvm-devel`` ports includes LLDB as a default option, this gives FreeBSD CoreCLR LLDB debug support for real.

Fix #797.

Related: #1154, #940, #1131. Thanks for the help on these threads.

Thanks to @brooksdavis for helping bring LLDB as a default option in ``devel/llvm37`` and ``devel/llvm-devel``.

### CODE REVIEW

**Comment by jasonwilliams200OK** [comment]
Thanks! :+1: 

FYI FreeBSD guys: @josteink, @saper, @ghuntley, @janhenke, @khellang & @ajensenwaud

**Comment by josteink** [comment]
Apart from my nits and the other minor inaccuracies observed, this looks like a very solid upgrade to our current build-documentation. Good job :)

**Comment by tony** [comment]
Hold up. This needs some corrections to it.  There are a few things to address:

- Unlike with ``clang35`` and ``clang36``, There isn't a ``lang/clang37`` package.  ``/usr/local/bin/clang37`` comes from ``devel/llvm37``. @brooksdavis: Can you help clarify how this works? Will there be a clang37 package in the future that pulls llvm37 with lldb? Up to now we've been giving instruction to download clang, which would grab llvm as a dep.
- The binary packages may not be up yet, in lieu of that, I'm going to give instruction on how to compile and install ``devel/llvm37`` and ``devel/llvm-devel`` from ports. Is that ok?

What do you think? @josteink, @saper, @ghuntley, @janhenke, @khellang & @ajensenwaud

**Comment by tony** [comment]
@josteink Got the highlighting stuff last commit.  Working on the other things.

We do have a few bigger issues  to tackle. See https://github.com/dotnet/coreclr/pull/1336#commitcomment-12503169

> There isn't currently a clang37 package. I'm debating if I should make one. It would just depend on llvm37 since it's now impractical to build clang or lldb separately from llvm.

Yep, we need the docs to clarify that.  ``devel/llvm36`` won't build clang, but ``devel/llvm37`` will build clang.  Is that accurate?

Team: Should we still give instructions for building with clang35 and clang36, but clarify the package difference with llvm37 and onward?

**Comment by josteink** [comment]
Building any part of that whole LLVM thing is usually a time and resource-sink which for most people is best avoided. If we can find a simple and working approach which is all based on prebuilt packages, that's the approach I would prefer.

If we need more packages defined/made to keep our package-dependency-story consistent between LLVM/clang-versions, I'm could almost be in favour of getting those packages in, even if it meant cheating.

But if we had to resort to that, it instead cause some internal inconsistencies in the FreeBSD package-tree and that's probably worse, with a much longer tail to keep happy than our teenie weenie documentation here.

If we can avoid cheating, I'm all for new packages though.

But for now, until we have some packages which behaves consistently across versions, we can take the cost of a little added documentation on our part.

**Comment by tony** [comment]
No rush on this. We've waited this long, let's get the docs right. :)

> Building any part of that whole LLVM thing is usually a time and resource-sink which for most people is best avoided. If we can find a simple and working approach which is all based on prebuilt packages, that's the approach I would prefer.

Agreed on that.

I guess the real question is when the new ``devel/llvm37`` and ``devel/llvm-devel`` with LLDB will show up for users in ``pkg``?

> If we need more packages defined/made to keep our package-dependency-story consistent between LLVM/clang-versions, I'm could almost be in favour of getting those packages in, even if it meant cheating.

I'm trying to find a way to reword it temporarily without adding confusion. But it kind of is weird from a user perspective because it does it a flip:

- clang35, clang36 will grab llvm35, llvm36 for you, respectively.
- llvm37 includes clang37, so there's no clang37 package.

@brooksdavis (I guess stuff like this is better for outside this issue but since we're here) Is it feasible to make it so devel/llvm35 + lang/clang35 and devel/llvm36 + lang/clang36 follow the new port format llvm37 uses? Since they're grabbing devel/llvm35 and devel/llvm36 anyway.  I like the "new way" of building llvm introduced in devel/llvm37.

**Comment by saper** [comment]
> I guess the real question is when the new ``devel/llvm37`` and
> ``devel/llvm-devel`` with LLDB will show up for users in ``pkg``?

I just did:

$ sudo pkg search llvm37
llvm37-3.7.0.r1

So it seems to be here, now :)

**Comment by tony** [comment]
> So it seems to be here, now :)

FreeBSD is going to have .NET coreclr power.

**Comment by brooksdavis** [comment]
The current state is that for clang/llvm 3.5 and 3.6, you get llvm from the devel/llvm3[56] ports and clang from the lang/clang3[56] ports.  With 3.7 and the -devel port, I've changed things around so everything happens in devel/llvm* ports.  In the case of -devel I converted the clang port to a meta port that just depend on llvm-devel and installs nothing.  If symmetry is helpful I could add a clang37 port, but it would do nothing other than making llvm37 easier to discover.

I don't currently have any plans to update the old llvm ports.  That would be quite a lot of work for little benefit and I'm not sure how solid the cmake support is in 3.5 and 3.6.  I'm also still trying to shake out some issues in the new ports.

Bottom line: Unless there are regressions in 3.7, I'd probably switch the instructions to point to it and drop the other bits, but if you want to support older versions I could add a clang37 port to make things simpler.

**Comment by janhenke** [comment]
Many things have already be said here. I agree to @josteink. In particular we should stick with ```pkg```, we really should avoid the can of worms to include compiling LLVM/Clang yourself.

The only other thing, you will have to squash your commits into one before it will get merged.

**Comment by tony** [comment]
Team,

At the advice of @brooksdavis, should we remove references to clang35, clang36, llvm-devel and just mention the llvm37 package?

Once we have that clarified I'll adjust accordingly, proofread and squash.

**Comment by josteink** [comment]
As long as it works, I'm completely OK with that.

It will probably simplify the instructions which is a definitive bonus.

To avoid confusion, you can add some notes (in parenthesis?) that things should work with older versions of clang too, so that people who has specific versions for specific reasons don't feel they need to break their existing setup.

But if others decides that this is overkill or not really needed, I'm OK with that too.

**Comment by janhenke** [comment]
I don't think that works right now. Clang 3.5 is still the default version, since that is the version included in Ubuntu 14.04 LTS.

It was made clear that the default will not be changed until there is a new Ubuntu LTS release. So we need to reflect that in the instructions.

**Comment by mmitche** [comment]
@dotnet-bot test this please

**Comment by tony** [comment]
Ok squashed, and nailed some other issues too.

To keep it simple, we just mention llvm37, and add a link to a section elaborating on Clang / LLVM at the bottom.

File: Documentation/building/freebsd-instructions.md [hunk 1]
```diff
@@ -19,43 +19,36 @@ Install the following packages for the toolchain:
...
-
 To install the packages you need:
 
-`janhenke@freebsd-frankfurt:~ % sudo pkg install bash cmake clang35 libunwind gettext`
+```sh
+janhenke@freebsd-frankfurt:~ % sudo pkg install bash cmake libunwind gettext llvm37
+```
 
-You now have all the required components.
+Want to user a lower version of Clang / LLVM? See [Note on Clang/LLVM versions](#note-on-clangllvm-versions).
```
Comment (janhenke) [inline Documentation/building/freebsd-instructions.md]: Can you rephrase that sentence? The style of speech does not fit the rest of the text and it is also not really what you expect in a technical documentation.

Comment (tony) [inline Documentation/building/freebsd-instructions.md]: No it's not. Thanks for catching that.

**Comment by josteink** [comment]
I can see all my nits are addressed. LGTM.

Good job!

**Comment by jkotas** [comment]
:+1:
